frontend/appflowy_flutter/packages/appflowy_editor/lib/src/render/rich_text/rich_text.dart

```diff
@@ -80,4 +88,95 @@ class _RichTextNodeWidgetState extends State<RichTextNodeWidget>
       ),
     );
   }
+
+  final List<OverlayEntry> _selectionOverlays = [];
+  final List<OverlayEntry> _cursorOverlays = [];
+
+  Selection? _cacheSelection;
+
+  @override
+  void setSelection(Selection? selection) {
+    if (_cacheSelection == selection) {
+      return;
+    }
+    _cacheSelection = selection;
+    _selectionOverlays
+      ..forEach((element) => element.remove())
+      ..clear();
+    _cursorOverlays
+      ..forEach((element) => element.remove())
+      ..clear();
+    if (selection == null) {
+      return;
+    }
+    selection = selection.normalized;
+    final path = widget.textNode.path;
+    if (path < selection.start.path || path > selection.end.path) {
+      return;
+    }
+    var newSelection = selection.copyWith();
+    if (selection.isSingle) {
+      if (path.equals(newSelection.start.path)) {
+        if (selection.isCollapsed) {
+          final rect = getCursorRectInPosition(newSelection.start);
+          if (rect != null) {
+            final entry = OverlayEntry(
+              builder: (context) {
+                return CursorWidget(
+                  layerLink: widget.textNode.layerLink,
+                  rect: rect,
+                  color: Colors.blue,
+                );
+              },
+            );
+            _cursorOverlays.add(entry);
+            Overlay.of(context)?.insertAll(_cursorOverlays);
+          }
+        } else {
+          final rects = getRectsInSelection(newSelection);
+          for (final rect in rects) {
+            final entry = OverlayEntry(
+              builder: (context) {
+                return SelectionWidget(
+                  layerLink: widget.textNode.layerLink,
+                  rect: rect,
+                  color: Colors.blue.withOpacity(0.3),
+                );
+              },
+            );
+            _selectionOverlays.add(entry);
+          }
+          Overlay.of(context)?.insertAll(_selectionOverlays);
+        }
+      }
+    } else {
+      if (path.equals(newSelection.start.path)) {
+        newSelection = newSelection.copyWith(end: end());
+      } else if (path.equals(newSelection.end.path)) {
+        newSelection = newSelection.copyWith(start: start());
+      } else {
+        newSelection = Selection(start: start(), end: end());
+      }
+      final rects = getRectsInSelection(newSelection);
+      for (final rect in rects) {
+        final entry = OverlayEntry(
+          builder: (context) {
+            return SelectionWidget(
+              layerLink: widget.textNode.layerLink,
+              rect: rect,
+              color: Colors.blue.withOpacity(0.3),
+            );
+          },
+        );
+        _selectionOverlays.add(entry);
+      }
+      Overlay.of(context)?.insertAll(_selectionOverlays);
+    }
+  }
```

Comment:
```suggestion
    if (!selection.isSingle) {
      if (path.equals(newSelection.start.path)) {
        newSelection = newSelection.copyWith(end: end());
      } else if (path.equals(newSelection.end.path)) {
        newSelection = newSelection.copyWith(start: start());
      } else {
        newSelection = Selection(start: start(), end: end());
      }
    }

    final rects = getRectsInSelection(newSelection);
    if (rect == null) {
      return;
    }

    for (final rect in rects) {
      final entry = OverlayEntry(
        builder: (context) {
          return SelectionWidget(
            layerLink: widget.textNode.layerLink,
            rect: rect,
            color: Colors.blue.withOpacity(0.3),
          );
        },
      );
      _selectionOverlays.add(entry);
    }

    Overlay.of(context)?.insertAll(_selectionOverlays); 
  }
```
Maybe something like this?

Reply:
Hey, @Xazin. This PR is a draft. Just to validate the logic. I will format and refactor all the code once this logic is validated without any issues.